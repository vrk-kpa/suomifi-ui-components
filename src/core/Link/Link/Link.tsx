import React, { forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { IconChevronRight } from 'suomifi-icons';
import { LinkStyles } from '../Link/Link.baseStyles';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import { HtmlA } from '../../../reset';
import {
  BaseLinkProps,
  baseClassName,
  linkClassNames,
} from '../BaseLink/BaseLink';
import { filterDuplicateKeys } from '../../../utils/common/common';

export interface LinkProps extends BaseLinkProps, MarginProps {
  /** Ref is forwarded to the anchor element. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLAnchorElement>;
}

const StyledLink = styled(
  ({
    asProp,
    className,
    smallScreen,
    theme,
    globalMargins,
    variant = 'default',
    children,
    underline = 'hover',
    ...rest
  }: LinkProps & SuomifiThemeProp & GlobalMarginProps) => {
    const [_marginProps, passProps] = separateMarginProps(rest);

    return (
      <HtmlA
        {...passProps}
        className={classnames(baseClassName, className, {
          [linkClassNames.linkUnderline]: underline === 'initial',
          [linkClassNames.accent]: variant === 'accent',
          [linkClassNames.small]: smallScreen,
        })}
        as={asProp}
        style={{ ...passProps?.style }}
      >
        {variant === 'accent' && (
          <IconChevronRight
            color={theme.colors.accentBase}
            className={linkClassNames.accentIcon}
          />
        )}
        {children}
      </HtmlA>
    );
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.button,
      marginProps,
    );
    return LinkStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const Link = forwardRef(
  (props: LinkProps, ref: React.Ref<HTMLAnchorElement>) => (
    <SpacingConsumer>
      {({ margins }) => (
        <SuomifiThemeConsumer>
          {({ suomifiTheme }) => (
            <StyledLink
              theme={suomifiTheme}
              globalMargins={margins}
              forwardedRef={ref}
              {...props}
            />
          )}
        </SuomifiThemeConsumer>
      )}
    </SpacingConsumer>
  ),
);

Link.displayName = 'Link';
export { Link };
