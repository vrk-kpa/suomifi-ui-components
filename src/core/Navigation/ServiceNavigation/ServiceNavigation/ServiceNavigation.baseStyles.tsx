import { font } from '../../../theme/reset';
import { css } from 'styled-components';
import { SuomifiTheme } from '../../../theme';
import { MarginProps, buildSpacingCSS } from '../../../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${font(theme)('bodyText')}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  &.fi-service-navigation {
    &--small-screen {
      background: ${theme.colors.highlightLight3};
      .fi-service-navigation-item,
      .fi-service-navigation-item--selected {
        border-bottom: 1px solid ${theme.colors.whiteBase};
      }
      .fi-service-navigation_expand-button {
        ${font(theme)('heading4')}
        position: relative;
        width: 100%;
        display: flex;
        justify-content: space-between;
        align-items: center;
        padding: ${theme.spacing.m};
        border: 1px solid ${theme.colors.highlightLight2};

        .fi-service-navigation_expand-button_text-wrapper {
          display: flex;
          align-items: center;
          justify-content: center;
        }

        &:focus {
          ${theme.focuses.highContrastFocus} /* For high contrast mode */
          &:after {
            ${theme.focuses.absoluteFocus}
          }
        }

        .fi-icon {
          height: 25px;
          width: 25px;
        }
      }
    }
    .fi-service-navigation_list {
      list-style-type: none;
      margin: 0;
      padding: 0;
    }
  }
`;
