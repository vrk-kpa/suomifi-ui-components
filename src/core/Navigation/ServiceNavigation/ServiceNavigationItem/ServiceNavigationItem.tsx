import React, { ReactNode } from 'react';
import classnames from 'classnames';
import { HtmlLi } from '../../../../reset';
import { SuomifiThemeConsumer, SuomifiThemeProp } from '../../../theme';
import { baseStyles } from './ServiceNavigationItem.baseStyles';
import { styled } from 'styled-components';

export interface ServiceNavigationItemProps {
  /** CSS class for custom styles */
  className?: string;
  /** Use the polymorphic `<RouterLink>` component as child to get intended CSS styling */
  children: ReactNode;
  /** Shows this item as the selected one */
  selected?: boolean;
  /** Disables the item */
  disabled?: boolean;
}

const baseClassName = 'fi-service-navigation-item';
const selectedClassName = `${baseClassName}--selected`;
const disabledClassName = `${baseClassName}--disabled`;

const BaseServiceNavigationItem = ({
  selected,
  className,
  children,
  disabled,
  ...passProps
}: ServiceNavigationItemProps) => (
  <HtmlLi
    className={classnames(className, {
      [baseClassName]: !selected,
      [selectedClassName]: selected,
      [disabledClassName]: disabled,
    })}
    aria-disabled={disabled}
    {...passProps}
  >
    {children}
  </HtmlLi>
);

const StyledServiceNavigationItem = styled(
  (props: ServiceNavigationItemProps & SuomifiThemeProp) => {
    const { theme, ...passProps } = props;
    return <BaseServiceNavigationItem {...passProps} />;
  },
)`
  ${({ theme }) => baseStyles(theme)}
`;

const ServiceNavigationItem = (props: ServiceNavigationItemProps) => (
  <SuomifiThemeConsumer>
    {({ suomifiTheme }) => (
      <StyledServiceNavigationItem theme={suomifiTheme} {...props} />
    )}
  </SuomifiThemeConsumer>
);

ServiceNavigationItem.displayName = 'ServiceNavigationItem';
export { ServiceNavigationItem };
