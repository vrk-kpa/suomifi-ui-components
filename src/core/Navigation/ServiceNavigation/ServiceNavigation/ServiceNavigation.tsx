import React, { forwardRef, ReactNode, useState } from 'react';
import {
  HtmlButton,
  HtmlDiv,
  HtmlDivProps,
  HtmlNav,
  HtmlSpan,
  HtmlUl,
} from '../../../../reset';
import styled from 'styled-components';
import { IconChevronDown, IconChevronRight } from 'suomifi-icons';
import {
  SpacingConsumer,
  SuomifiThemeConsumer,
  SuomifiThemeProp,
} from '../../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../../theme/utils/spacing';
import { getConditionalAriaProp } from '../../../../utils/aria';
import { baseStyles } from './ServiceNavigation.baseStyles';
import classnames from 'classnames';
import { filterDuplicateKeys } from '../../../../utils/common/common';

export interface ServiceNavigationProps extends MarginProps, HtmlDivProps {
  /** Use `<ServiceNavigationItem>` components as children */
  children: ReactNode;
  /** Name for the navigation element. Don't use the word "navigation" since it will be read by screen readers regardless. */
  'aria-label': string;
  /** HTML id attribute for the `<nav>` element. Typically used together with a `<SkipLink>` to quickly move to the nav */
  id?: string;
  /**
   * Component variant. `'smallScreen'` allows the navigation to be collapsed/expanded with a toggle button
   * @default normal
   */
  variant?: 'default' | 'smallScreen';
  /**
   * Initial expand status of the menu. Only applies to the smallScreen variant
   * @default true
   */
  initiallyExpanded?: boolean;
  /** When using the smallScreen variant, this is the text showed in the expander button */
  smallScreenExpandButtonText?: string | ReactNode;
  /** CSS class for custom styles */
  className?: string;
  /** Ref is forwarded to the nav element. Alternative for React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLElement>;
}

const baseClassName = 'fi-service-navigation';
const smallScreenClassName = `${baseClassName}--small-screen`;
const listClassName = `${baseClassName}_list`;
const smallScreenExpandButtonClassName = `${baseClassName}_expand-button`;
const buttonTextClassName = `${baseClassName}_expand-button_text-wrapper`;

const BaseServiceNavigation = ({
  children,
  'aria-label': ariaLabel,
  id,
  variant,
  initiallyExpanded,
  smallScreenExpandButtonText,
  className,
  forwardedRef,
  style,
  ...rest
}: ServiceNavigationProps) => {
  const [_marginProps, passProps] = separateMarginProps(rest);

  const initiallyExpandedValue =
    initiallyExpanded !== undefined ? initiallyExpanded : true;
  const [smallScreenNavOpen, setSmallScreenNavOpen] = useState(
    initiallyExpandedValue,
  );

  return (
    <HtmlDiv
      className={classnames(baseClassName, className, {
        [smallScreenClassName]: variant === 'smallScreen',
      })}
      style={style}
      {...passProps}
    >
      {variant === 'smallScreen' && (
        <HtmlButton
          className={smallScreenExpandButtonClassName}
          onClick={() => setSmallScreenNavOpen(!smallScreenNavOpen)}
          aria-expanded={smallScreenNavOpen}
        >
          <HtmlSpan className={buttonTextClassName}>
            {smallScreenExpandButtonText}
          </HtmlSpan>
          {smallScreenNavOpen ? <IconChevronDown /> : <IconChevronRight />}
        </HtmlButton>
      )}
      {((variant === 'smallScreen' && smallScreenNavOpen) ||
        variant !== 'smallScreen') && (
        <HtmlNav
          forwardedRef={forwardedRef}
          id={id}
          {...getConditionalAriaProp('aria-label', [ariaLabel])}
        >
          <HtmlUl className={listClassName} role="list">
            {children}
          </HtmlUl>
        </HtmlNav>
      )}
    </HtmlDiv>
  );
};

const StyledServiceNavigation = styled(
  (props: ServiceNavigationProps & SuomifiThemeProp & GlobalMarginProps) => {
    const { theme, globalMargins, ...passProps } = props;
    return <BaseServiceNavigation {...passProps} />;
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.serviceNavigation,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const ServiceNavigation = forwardRef(
  (props: ServiceNavigationProps, ref: React.Ref<HTMLElement>) => (
    <SpacingConsumer>
      {({ margins }) => (
        <SuomifiThemeConsumer>
          {({ suomifiTheme }) => (
            <StyledServiceNavigation
              theme={suomifiTheme}
              globalMargins={margins}
              forwardedRef={ref}
              {...props}
            />
          )}
        </SuomifiThemeConsumer>
      )}
    </SpacingConsumer>
  ),
);

ServiceNavigation.displayName = 'ServiceNavigation';
export { ServiceNavigation };
