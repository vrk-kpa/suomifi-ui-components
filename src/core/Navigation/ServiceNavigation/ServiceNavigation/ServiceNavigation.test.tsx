import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../../utils/test';

import { ServiceNavigation } from './ServiceNavigation';
import { ServiceNavigationItem } from '../ServiceNavigationItem/ServiceNavigationItem';
import { ExternalLink, RouterLink } from '../../../Link';
import { StaticChip } from '../../../Chip';

interface TestProps {
  children: string;
  onClick?: () => void;
  role?: string;
  tabIndex?: number;
}

const Comp = ({ children, ...passProps }: TestProps) => (
  <div {...passProps}>{children}</div>
);

const TestServiceNavigation = (
  <ServiceNavigation aria-label="Test">
    <ServiceNavigationItem>
      <RouterLink href="/" aria-label="Inbox. 16 unread messages.">
        Inbox
        <StaticChip style={{ marginLeft: '15px' }} aria-hidden>
          16
        </StaticChip>
      </RouterLink>
    </ServiceNavigationItem>
    <ServiceNavigationItem>
      <RouterLink asComponent={ExternalLink} href="https://suomi.fi" hideIcon>
        Sent
      </RouterLink>
    </ServiceNavigationItem>
    <ServiceNavigationItem selected>
      <RouterLink
        asComponent={ExternalLink}
        href="https://www.suomi.fi"
        hideIcon
        aria-current="page"
      >
        New Message
      </RouterLink>
    </ServiceNavigationItem>
    <ServiceNavigationItem>
      <RouterLink
        asComponent={ExternalLink}
        href="https://www.suomi.fi"
        hideIcon
      >
        Drafts
      </RouterLink>
    </ServiceNavigationItem>
    <ServiceNavigationItem>
      <RouterLink
        asComponent={Comp}
        onClick={() => console.log('Nav item clicked')}
      >
        Settings
      </RouterLink>
    </ServiceNavigationItem>
    <ServiceNavigationItem disabled>
      <RouterLink>Devices</RouterLink>
    </ServiceNavigationItem>
  </ServiceNavigation>
);

it('should match snapshot', () => {
  const NavRendered = render(TestServiceNavigation);
  const { container } = NavRendered;
  expect(container.firstChild).toMatchSnapshot();
});

test(
  'should not have basic accessibility issues',
  axeTest(TestServiceNavigation),
);

describe('margin', () => {
  it('should have margin style from margin prop', () => {
    const { container } = render(
      <ServiceNavigation aria-label="" margin="xs">
        <ServiceNavigationItem>Test</ServiceNavigationItem>
      </ServiceNavigation>,
    );
    expect(container.firstChild).toHaveStyle('margin: 10px');
  });

  it('should have margin style overridden by style prop', async () => {
    const { container } = render(
      <ServiceNavigation aria-label="" margin="xs" style={{ margin: 2 }}>
        <ServiceNavigationItem>Test</ServiceNavigationItem>
      </ServiceNavigation>,
    );
    expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
  });
});
