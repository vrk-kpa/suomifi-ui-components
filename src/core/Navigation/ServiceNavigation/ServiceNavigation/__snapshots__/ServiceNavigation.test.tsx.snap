// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.c6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c6::before,
.c6::after {
  box-sizing: border-box;
}

.c6:hover,
.c6:active,
.c6:focus,
.c6:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: list-item;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c8 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c8::before,
.c8::after {
  box-sizing: border-box;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  list-style-type: decimal;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  padding-left: 40px;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c1 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c1.fi-service-navigation--small-screen {
  background: hsl(212,63%,95%);
}

.c1.fi-service-navigation--small-screen .fi-service-navigation-item,
.c1.fi-service-navigation--small-screen .fi-service-navigation-item--selected {
  border-bottom: 1px solid hsl(0,0%,100%);
}

.c1.fi-service-navigation--small-screen .fi-service-navigation_expand-button {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 24px;
  font-weight: 600;
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 20px;
  border: 1px solid hsl(212,63%,90%);
}

.c1.fi-service-navigation--small-screen .fi-service-navigation_expand-button .fi-service-navigation_expand-button_text-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1.fi-service-navigation--small-screen .fi-service-navigation_expand-button:focus {
  outline: 3px solid transparent;
}

.c1.fi-service-navigation--small-screen .fi-service-navigation_expand-button:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1.fi-service-navigation--small-screen .fi-service-navigation_expand-button .fi-icon {
  height: 25px;
  width: 25px;
}

.c1.fi-service-navigation .fi-service-navigation_list {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.c5 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
}

.c5.fi-service-navigation-item {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c5.fi-service-navigation-item:hover,
.c5.fi-service-navigation-item:active {
  background: hsl(212,63%,95%);
  cursor: pointer;
  border-left: 4px solid hsl(212,63%,45%);
}

.c5.fi-service-navigation-item:hover .fi-link--router,
.c5.fi-service-navigation-item:active .fi-link--router {
  padding-left: calc(20px - 4px);
}

.c5.fi-service-navigation-item:focus-within {
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  outline: 0;
  z-index: 9999;
}

.c5.fi-service-navigation-item .fi-link--router {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: hsl(0,0%,13%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 10px 20px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  cursor: pointer;
  background: inherit;
  border: none;
}

.c5.fi-service-navigation-item .fi-link--router:focus {
  border: none;
  box-shadow: none;
  color: hsl(0,0%,13%);
  outline: 3px solid transparent;
}

.c5.fi-service-navigation-item .fi-link--router:hover,
.c5.fi-service-navigation-item .fi-link--router:active,
.c5.fi-service-navigation-item .fi-link--router:visited {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: hsl(0,0%,13%);
}

.c5.fi-service-navigation-item--disabled:hover,
.c5.fi-service-navigation-item--disabled:active {
  background: transparent;
  cursor: not-allowed;
  border-left: none;
}

.c5.fi-service-navigation-item--disabled:hover .fi-link--router,
.c5.fi-service-navigation-item--disabled:active .fi-link--router {
  padding-left: 20px;
}

.c5.fi-service-navigation-item--disabled .fi-link--router {
  color: hsl(202,7%,67%);
  cursor: not-allowed;
  pointer-events: none;
}

.c5.fi-service-navigation-item--disabled .fi-link--router:hover,
.c5.fi-service-navigation-item--disabled .fi-link--router:active,
.c5.fi-service-navigation-item--disabled .fi-link--router:visited {
  color: hsl(202,7%,67%);
}

.c5.fi-service-navigation-item--selected {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: hsl(212,63%,95%);
  border-left: 4px solid hsl(212,63%,45%);
  position: relative;
}

.c5.fi-service-navigation-item--selected:focus {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c5.fi-service-navigation-item--selected .fi-link--router {
  font-weight: bold;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: hsl(0,0%,13%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 10px 25px;
  padding-left: calc(20px - 4px);
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5.fi-service-navigation-item--selected .fi-link--router:hover,
.c5.fi-service-navigation-item--selected .fi-link--router:active,
.c5.fi-service-navigation-item--selected .fi-link--router:visited {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: hsl(0,0%,13%);
}

.c11 {
  position: absolute;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

.c10 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c10.fi-link {
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c10.fi-link:hover,
.c10.fi-link:active,
.c10.fi-link:focus,
.c10.fi-link:focus-within {
  color: hsl(212,63%,45%);
}

.c10.fi-link:focus,
.c10.fi-link:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c10.fi-link:focus {
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  outline: 0;
  outline: 3px solid transparent;
}

.c10.fi-link:hover,
.c10.fi-link:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c10.fi-link:visited {
  color: hsl(284,36%,45%);
}

.c10.fi-link--initial-underline {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c10.fi-link--initial-underline:focus,
.c10.fi-link--initial-underline:focus-within {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c10.fi-link--initial-underline:hover,
.c10.fi-link--initial-underline:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c10.fi-link.fi-link--accent .fi-link--accent_icon {
  font-size: 16px;
  margin-left: -3px;
  margin-right: 2px;
  -webkit-transform: translateY(0.15em);
  -ms-transform: translateY(0.15em);
  transform: translateY(0.15em);
}

.c10.fi-link.fi-link--small {
  font-size: 16px;
}

.c10 .fi-link_icon {
  margin: 0;
  padding-left: 4px;
  font-size: 16px;
  box-sizing: content-box;
}

.c7 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c7.fi-link {
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link:hover,
.c7.fi-link:active,
.c7.fi-link:focus,
.c7.fi-link:focus-within {
  color: hsl(212,63%,45%);
}

.c7.fi-link:focus,
.c7.fi-link:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link:focus {
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  outline: 0;
  outline: 3px solid transparent;
}

.c7.fi-link:hover,
.c7.fi-link:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link:visited {
  color: hsl(284,36%,45%);
}

.c7.fi-link--initial-underline {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link--initial-underline:focus,
.c7.fi-link--initial-underline:focus-within {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link--initial-underline:hover,
.c7.fi-link--initial-underline:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link.fi-link--accent .fi-link--accent_icon {
  font-size: 16px;
  margin-left: -3px;
  margin-right: 2px;
  -webkit-transform: translateY(0.15em);
  -ms-transform: translateY(0.15em);
  transform: translateY(0.15em);
}

.c7.fi-link.fi-link--small {
  font-size: 16px;
}

.c9 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c9.fi-chip {
  border-radius: 14px;
  padding: 2px 10px;
  color: hsl(0,0%,100%);
  background: hsl(212,63%,45%);
  max-height: 100%;
  display: inline-block;
}

.c9.fi-chip:focus {
  outline: 0;
  position: relative;
}

.c9.fi-chip:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  border-radius: 16px;
}

.c9.fi-chip .fi-chip--content {
  display: block;
  word-break: break-word;
  overflow: hidden;
  line-height: 1.5em;
}

.c9.fi-chip--disabled.fi-chip {
  cursor: not-allowed;
  background: hsl(202,7%,67%);
}

.c9.fi-chip--disabled.fi-chip:hover,
.c9.fi-chip--disabled.fi-chip:active {
  background: hsl(202,7%,67%);
}

@media screen and (forced-colors:active) {
  .c9.fi-chip--disabled.fi-chip {
    color: GrayText;
  }
}

<div
  class="c0 fi-service-navigation c1"
>
  <nav
    aria-label="Test"
    class="c2"
  >
    <ul
      class="c3 fi-service-navigation_list"
      role="list"
    >
      <li
        class="c4 c5 fi-service-navigation-item"
      >
        <a
          aria-label="Inbox. 16 unread messages."
          class="c6 fi-link fi-link--router c7"
          href="/"
        >
          Inbox
          <span
            aria-hidden="true"
            class="c8 fi-chip c9"
            style="margin-left: 15px;"
          >
            <span
              class="c8 fi-chip--content"
            >
              16
            </span>
          </span>
        </a>
      </li>
      <li
        class="c4 c5 fi-service-navigation-item"
      >
        <a
          class="c6 fi-link c10 fi-link fi-link--router c7 fi-link--external"
          href="https://suomi.fi"
          rel="noopener"
          target="_blank"
        >
          Sent
          <span
            class="c8 c11 fi-visually-hidden"
          />
        </a>
      </li>
      <li
        class="c4 c5 fi-service-navigation-item--selected"
      >
        <a
          aria-current="page"
          class="c6 fi-link c10 fi-link fi-link--router c7 fi-link--external"
          href="https://www.suomi.fi"
          rel="noopener"
          target="_blank"
        >
          New Message
          <span
            class="c8 c11 fi-visually-hidden"
          />
        </a>
      </li>
      <li
        class="c4 c5 fi-service-navigation-item"
      >
        <a
          class="c6 fi-link c10 fi-link fi-link--router c7 fi-link--external"
          href="https://www.suomi.fi"
          rel="noopener"
          target="_blank"
        >
          Drafts
          <span
            class="c8 c11 fi-visually-hidden"
          />
        </a>
      </li>
      <li
        class="c4 c5 fi-service-navigation-item"
      >
        <div
          class="fi-link fi-link--router c7"
        >
          Settings
        </div>
      </li>
      <li
        aria-disabled="true"
        class="c4 c5 fi-service-navigation-item fi-service-navigation-item--disabled"
      >
        <a
          class="c6 fi-link fi-link--router c7"
        >
          Devices
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
